### PR TITLE
Add unit-tests for ApplicationRollout related packages

### PR DESCRIPTION
**Pull request by zzxwill**
Part I: pkg/controller/common/rollout

### CODE REVIEW

**Comment by zzxwill** [comment]
@ryanzhang-oss I slightly modified some code, please help review it:)

File: pkg/controller/common/rollout/workloads/deployment_controller.go [hunk 1]
```diff
@@ -73,15 +73,19 @@ func (c *DeploymentController) VerifySpec(ctx context.Context) (bool, error) {
 	defer func() {
 		if verifyErr != nil {
 			klog.Error(verifyErr)
-			c.recorder.Event(c.parentController, event.Warning("VerifyFailed", verifyErr))
+			if c.recorder != nil && c.parentController != nil {
```
Comment (ryanzhang-oss) [inline pkg/controller/common/rollout/workloads/deployment_controller.go:76]: when will the recorder be nil?

File: pkg/controller/common/rollout/workloads/deployment_controller.go [hunk 2]
```diff
@@ -73,15 +73,19 @@ func (c *DeploymentController) VerifySpec(ctx context.Context) (bool, error) {
...
+			}
 		}
 	}()
 
 	// check if the rollout spec is compatible with the current state
 	targetTotalReplicas, verifyErr := c.calculateTargetTotalSize(ctx)
 	if verifyErr != nil {
 		// do not fail the rollout just because we can't get the resource
-		c.rolloutStatus.RolloutRetry(verifyErr.Error())
+		if c.rolloutSpec != nil {
```
Comment (ryanzhang-oss) [inline pkg/controller/common/rollout/workloads/deployment_controller.go:86]: when will the spec be nil?

File: pkg/controller/common/rollout/workloads/deployment_controller.go (hunk 1 above)
Comment (ryanzhang-oss) [reply pkg/controller/common/rollout/workloads/deployment_controller.go:76]: btw, the deployment rollout is not really finished, it doesn't cover many corner cases (i.e. revert in the middle of rollout)

Comment (zzxwill) [reply pkg/controller/common/rollout/workloads/deployment_controller.go:76]: @ryanzhang-oss From the normal logic, application rollout plan controller --> rollout plan controller --> deployment controller, the recorder won't be nil.
But from the filed type, it's possible.

>btw, the deployment rollout is not really finished, it doesn't cover many corner cases (i.e. revert in the middle of rollout)

Got it, I will focus on adding unit-tests on Cloneset.

File: pkg/controller/common/rollout/workloads/deployment_controller.go (hunk 2 above)
Comment (zzxwill) [reply pkg/controller/common/rollout/workloads/deployment_controller.go:86]: Same reason as above. Should I remove the check nil logic?

Comment (ryanzhang-oss) [reply pkg/controller/common/rollout/workloads/deployment_controller.go:86]: yeah, we should not check (and allow) conditions that should not happen.

File: pkg/controller/common/rollout/workloads/cloneset_rollout_controller.go [hunk 3]
```diff
@@ -188,6 +188,11 @@ func (c *CloneSetRolloutController) CheckOneBatchPods(ctx context.Context) (bool
 	newPodTarget := calculateNewBatchTarget(c.rolloutSpec, 0, int(cloneSetSize), int(c.rolloutStatus.CurrentBatch))
 	// get the number of ready pod from cloneset
 	readyPodCount := int(c.cloneSet.Status.UpdatedReadyReplicas)
+	if len(c.rolloutSpec.RolloutBatches) <= int(c.rolloutStatus.CurrentBatch) {
+		klog.Info("somehow, currentBatch number exceeded the rolloutBatches spec.", "current batch",
+			c.rolloutStatus.CurrentBatch)
+		return false, nil
+	}
```
Comment (zzxwill) [inline pkg/controller/common/rollout/workloads/cloneset_rollout_controller.go:195]: @ryanzhang-oss I kept this checking and removed all other nil checking.

File: pkg/controller/common/rollout/workloads/deployment_controller.go (hunk 2 above)
Comment (zzxwill) [reply pkg/controller/common/rollout/workloads/deployment_controller.go]: Okay, removed.
